cmd/soroban-rpc/internal/test/upgrade_test.go

```diff
@@ -0,0 +1,128 @@
+		IncrementSequenceNum: true,
+		Operations: []txnbuild.Operation{
+			createInstallContractCodeOperation(account.AccountID, helloWorldContract),
+		},
+		BaseFee: txnbuild.MinBaseFee,
+		Preconditions: txnbuild.Preconditions{
+			TimeBounds: txnbuild.NewInfiniteTimeout(),
+		},
+	})
+
+	tx, err := txnbuild.NewTransaction(params)
+	assert.NoError(t, err)
+	sendSuccessfulTransaction(t, client, sourceAccount, tx)
+
+	// Upgrade to protocol 21 and re-upload the contract, which should cause a caching of the contract
+	// estimations
+	test.UpgradeProtocol(21)
+	// Wait for the ledger to advance, so that the simulation library passes the right protocol
+	rpcDB := test.daemon.GetDB()
+	initialLedgerSequence, err := db.NewLedgerEntryReader(rpcDB).GetLatestLedgerSequence(context.Background())
+	require.NoError(t, err)
+	newLedgerSequence := uint32(0)
+	for i := 0; i < 60; i++ {
+		time.Sleep(time.Second)
+		newLedgerSequence, err = db.NewLedgerEntryReader(rpcDB).GetLatestLedgerSequence(context.Background())
+		require.NoError(t, err)
+		if newLedgerSequence > initialLedgerSequence {
+			break
+		}
+	}
+	require.Greater(t, newLedgerSequence, initialLedgerSequence, "rpc didn't start ingesting protocol 21")
+
+	params = preflightTransactionParams(t, client, txnbuild.TransactionParams{
+		SourceAccount:        &account,
+		IncrementSequenceNum: true,
+		Operations: []txnbuild.Operation{
+			createInstallContractCodeOperation(account.AccountID, helloWorldContract),
+		},
+		BaseFee: txnbuild.MinBaseFee,
+		Preconditions: txnbuild.Preconditions{
+			TimeBounds: txnbuild.NewInfiniteTimeout(),
+		},
+	})
+
+	tx, err = txnbuild.NewTransaction(params)
+	assert.NoError(t, err)
+	sendSuccessfulTransaction(t, client, sourceAccount, tx)
+
+	params = preflightTransactionParams(t, client, txnbuild.TransactionParams{
+		SourceAccount:        &account,
+		IncrementSequenceNum: true,
+		Operations: []txnbuild.Operation{
+			createCreateContractOperation(address, helloWorldContract),
+		},
+		BaseFee: txnbuild.MinBaseFee,
+		Preconditions: txnbuild.Preconditions{
+			TimeBounds: txnbuild.NewInfiniteTimeout(),
+		},
+	})
+
+	tx, err = txnbuild.NewTransaction(params)
+	assert.NoError(t, err)
+	sendSuccessfulTransaction(t, client, sourceAccount, tx)
+
+	contractID := getContractID(t, address, testSalt, StandaloneNetworkPassphrase)
+	contractFnParameterSym := xdr.ScSymbol("world")
+	params = preflightTransactionParams(t, client, txnbuild.TransactionParams{
+		SourceAccount:        &account,
+		IncrementSequenceNum: true,
+		Operations: []txnbuild.Operation{
+			createInvokeHostOperation(
+				address,
+				contractID,
+				"hello",
+				xdr.ScVal{
+					Type: xdr.ScValTypeScvSymbol,
+					Sym:  &contractFnParameterSym,
+				},
+			),
+		},
+		BaseFee: txnbuild.MinBaseFee,
+		Preconditions: txnbuild.Preconditions{
+			TimeBounds: txnbuild.NewInfiniteTimeout(),
+		},
+	})
+	tx, err = txnbuild.NewTransaction(params)
+	assert.NoError(t, err)
+
+	assert.NoError(t, err)
```

Comment:
duplicate assertion on ln 124 and 126. 
is it worthwhile to check return value/SorobanData on the tx result to confirm invoke host fn response?

Reply:
Good catch